File: src/fit_clasif.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plot

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, confusion_matrix, accuracy_score
from xgboost import XGBClassifier
import pickle
import cdsw

# Split train-test
data = pd.read_csv('Bank_Loan_clean.csv')
train_set, test_set = train_test_split(data.drop(['Unnamed: 0','ID'], axis=1), test_size=0.8 , random_state=100)
y_train = train_set.pop('Personal Loan')
y_test = test_set.pop('Personal Loan')         

scaler = MinMaxScaler()
x_train = scaler.fit_transform(train_set)
x_test = scaler.transform(test_set)

# modelo
xgb_clf = XGBClassifier()
xgb_clf.fit(x_train, y_train.values)

pred = xgb_clf.predict(x_test)
# calculamos precisión y auc
confusion_matrix(y_test, pred)

auc = roc_auc_score(y_test, pred)
acc = accuracy_score(y_test, pred)

# Set metrics to track
#cdsw.track_metric('acc', acc)
#cdsw.track_metric('auc', auc)


# Output
filename = 'model.pkl'
with open(filename, 'wb') as model_file:
  pickle.dump(xgb_clf, model_file)
#cdsw.track_file(filename)

File: R/fit_clasif.r
# cargamos las librerías de visualización
if(!require(caret)){
    install.packages('caret')
    library(caret)
}
if(!require(xgboost)){
    install.packages('xgboost', dependencies=TRUE)
    library(xgboost)
}
if (!require("pROC")) {
  install.packages("pROC")
  library(pROC)
}
if (!require("rjson")) {
  install.packages("rjson")
  install.packages("jsonlite")
  library(rjson)
  library(jsonlite)
}

library (cdsw)

# Split train-test
data <- read.csv("Bank_Loan_Clean.csv")
num <- createDataPartition(data$Personal.Loan, p = .80, list = F)
train_set <- data[num,]
test_set <- data[-num,]
#train <- subset(data, select = -c(Personal.Loan))
#label <- subset(data, select = c(Personal.Loan))

set.seed(123)
train <- data.matrix(train_set[,c(-1,-10)])
output_vector = train_set[,'Personal.Loan'] == "1"

boost <- xgboost(data =train, 
                 label = output_vector,
                 nthread = 2,
                 nrounds = 2,
                 objective = "binary:logistic")


y_pred <- predict(boost, data.matrix(test_set[,c(-1,-10)]))

# si la probabilidad es mayor al 65%, será 1
pred <- integer(length((y_pred)))
pred[(y_pred >= 0.65)] <- 1

# Checking classification accuracy
cm <- table(pred, test_set$Personal.Loan)
cm

#calcula métricas de validación (AUC y precisión)
roc_obj <- roc(test_set$Personal.Loan, pred)
auc = auc(roc_obj)
acc = 100 * sum(diag(cm)) / sum(cm)
acc

# Define métricas para seguir en el workbench
#cdsw::track.metric("acc", acc)
#cdsw::track.metric("auc", auc)

# save model
filename = "model.pkl"
saveRDS(boost, file=filename)
#cdsw::track.file(filename)

write.csv(test_set, 'test.csv')
#test <- data.matrix(test_set[,c(-1,-10)])
#x <-toJSON(test, dataframe = c("columns"),matrix = c("rowmajor"))
#write_json(list("clients"=x), "test.json")
#result <- fromJSON("test.json")
